recitation_3/memorypointers.md

```diff
@@ -0,0 +1,282 @@
+```
+
+The `sizeof()` operator returns the number of bytes occupied by the array. In this case, `a` is an array of 10 `int` elements. The `sizeof(int)` is 4, therefore the `sizeof(a)` = 40.
+
+## Pointer Arithmetic ##   
+
+If you have a pointer, you can do basic arithmetic with it to address adjacent
+elements. All arithmetic is with respect to the type of element being addressed,
+so if you have an int pointer `int *p`, `p+1` points to the next int, which is 4
+bytes later. *Think in terms of elements, not in terms of bytes.*
+
+
+## Arrays and Pointers ## 
+
+Arrays and pointers behave very similarly, but it is important to know the differences!
+
+
+
+Given pointer `p` of type `T*` and integer `i`
+
+`*(p+i) == p[i]`
+
+In other words, this means that the `p[i]` is exactly the same thing as `*(p+i)`. Essentially, we can use the square brackets `[]` with a pointer to find the address of the `i`th element away from `p`. Then we can dereference the new address and voilá, we have the `i`th element itself!
+
+If this is sounding a lot like an array, that's because it basically is! If we set a pointer to the beginning of an array, we can use the exact same syntax to access the elements in the array that we want.
+
+
+```c
+int a[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+int *p = a; // p points to the first element of `a`, 0
+int *p = &a[0]; //p still points to 0
+
+*(p+5) == a[5]; // 5 == 5
+
+p+5 == a+5; // p+5 is a pointer to the 5th element of the array, so is a+5
+```
+
+This rule goes the other way too! Arrays, in cases of pointer arithmetic, operate as pointers to the first element.
+> i.e. Given an array `a` of type `T` elements </br>
+> a+1 == &a[0]+1
+
+So aren't pointers and arrays the same? **Wrong**! Here are the cases in which arrays do not act like pointers.
+
+**`sizeof` operator**
+
+As mentioned above, the size of an array is
+Note that as discussed above, `sizeof` is an operator, not a function. 
+Which means that for classic C `sizeof` is evaluated at *compile time*, so the
+value of the operator cannot be anything that depends on user input.
+
+**Array is a constant variable**
+
+```c
+int a[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+int *p = a; // p points to the first element of `a`, 0
+
+p++; // ok: `p` is just a pointer like any other, p now points to the next element in the array
+a++; // illegal (will throw an error): `a` is an array name, a constant variable
+```
+
+Unlike a pointer, though, an array is a constant variable. You cannot change its
+assignment after it has been created, it must point to the same chunk of memory.
+
+
+**Arrays cannot be passed into functions**
+
+Note that once you pass an array into a function, the array becomes a pointer to
+the first element, and loses all its array-ness. So within the scope where `int
+a[10]` was declared, `sizeof(a)` returns the number of bytes of the array `a`,
+ie 40. But if you pass `a` into a function as `arr`, then `sizeof(arr)` is NOT
+40, but 8, the size of a pointer.
```

Comment:
The fact that the size of ALL pointers is 8 bytes can be emphasized a bit more throughout the document because people generally forget that.